pkg/proxy/iptables/number_generated_rules_test.go

```diff
@@ -0,0 +1,206 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+import (
+	"fmt"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	iptablestest "k8s.io/kubernetes/pkg/util/iptables/testing"
+	netutils "k8s.io/utils/net"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// kube-proxy generates iptables rules to forward traffic from Services to Endpoints
+// kube-proxy uses iptables-restore to configure the rules atomically, however,
+// this has the downside that large number of rules take a long time to be processed,
+// causing disruption.
+// There are different parameters than influence the number of rules generated:
+// - ServiceType
+// - Number of Services
+// - Number of Endpoints per Service
+// This test will fail when the number of rules change, so the person
+// that is modifying the code can have feedback about the performance impact
+// on their changes. It also runs multiple number of rules test cases to check
+// if the number of rules grows linearly.
+func TestNumberIptablesRules(t *testing.T) {
+	testCases := []struct {
+		name                string
+		serviceType         v1.ServiceType
+		services            int
+		epPerService        int
+		expectedNatRules    int
+		expectedFilterRules int
+	}{
+		{
+			name:                "ClusterIP 10 Services 10 EndpointPerService",
+			serviceType:         v1.ServiceTypeClusterIP,
+			services:            10,
```

Comment:
I think it's useful to modulate params independently, and to start with 0.

0 services, 0 endpoints
1 service, 0 endpoints
1 service, 1 endpoint
1 service, 10 endpoints
10 service, 0 endpoints
10 service, 1 endpoint
10 service, 10 endpoints

Maybe repeat for 20, but I don't think you need to go to 100 or 1000?  Is that teaching us anything?